src/datasets/filesystem.py

```diff
@@ -0,0 +1,88 @@
+import fsspec
+import s3fs
```

Comment:
Since those are optional dependencies we should check if they actually exist in the users' env before loading them.
This can be done using
```python
import importlib

_has_s3fs = importlib.util.find_spec("s3fs") is not None
```

Then to be able to have the S3FileSystem imported only if s3fs is installed I would replace the filesystem.py by a `filesystem` module with an `__init__.py` that looks like
```python
import importlib

_has_s3fs = importlib.util.find_spec("s3fs") is not None

if _has_s3fs:
    from .s3filesystem import S3FileSystem
```

and `filesystem/s3filesystem.py` that can import `s3fs` since it's imported only if `_has_s3fs` is true.

Let me know what you think.

In the future this can be extended by checking other dependencies like gcsfs or fsspec itself.

Reply:
we have implemented this within our session together